[iOS] Sync the boot splash and the launch screen image scale modes

jkirsteins opened the pull request:
# What does this PR address?

Using iOS launch screens for the boot splash is not aligned with how Godot renders the boot splash.

On iOS the splash screen with launch images happens in two stages:
- show the launch screen (the export template allows modifying the scale mode - aspect fit, aspect fill etc.)
- afterwards Godot takes over, and renders the splash screen. **Godot does not take into account the iOS splash scale mode** so when it takes over, there is a visual tear as the splash changes dimensions and position.

This is partially addressing https://github.com/godotengine/godot/issues/81389 (though this PR does not address any compatibility issues with the legacy renderer)

# Change outline

- I added two methods in `Vector2` for fitting Vector2 inside another Vector2, either with aspect-fit-center or aspect-fit-cover modes (analogous to what the TextureRect exposes)
- I added Vector2 tests to verify the calculated areas
- I refactored `RendererCompositorRD` to allow the `OS` singleton to calculate the splash area. The iOS singleton will fetch the aspect mode from `Info.plist` and calculate a matching area. Others will keep existing behavior.

# Testing

Here's some visual examples of the effect.

## Landscape - export template specifies "scale to fill"

Before the changes:

https://github.com/user-attachments/assets/1b82b336-e04c-4a85-aed5-d863bcd4b734

After the changes:

https://github.com/user-attachments/assets/ac174b36-c712-4247-b431-481256eed518

## Portrait - export template specifies "scale"

Before the changes:

https://github.com/user-attachments/assets/74276062-2922-459d-a431-5488cdb42a28

After the changes:

https://github.com/user-attachments/assets/2d593f19-661c-4d68-9632-fdd4be92f334

jkirsteins committed:
[iOS] Sync the boot splash and the launch screen image scale modes

bruvzg commented on core/os/os.cpp:
```diff
@@ -222,6 +222,23 @@ uint64_t OS::get_embedded_pck_offset() const {
 }
 
+// Default boot screen rect scale mode is "Keep Aspect Centered"
+Rect2 OS::calculate_boot_screen_rect(const Size2 window_size, const Size2 imgrect_size) const {
```
```suggestion
Rect2 OS::calculate_boot_screen_rect(const Size2 &p_window_size, const Size2 &p_imgrect_size) const {
```

bruvzg commented on core/os/os.h:
```diff
@@ -161,6 +161,8 @@ class OS {
 	virtual void open_midi_inputs();
 	virtual void close_midi_inputs();
 
+	virtual Rect2 calculate_boot_screen_rect(const Size2 window_size, const Size2 imgrect_size) const;
```
```suggestion
    virtual Rect2 calculate_boot_screen_rect(const Size2 &p_window_size, const Size2 &p_imgrect_size) const;
```

bruvzg commented on misc/dist/ios_xcode/godot_ios/godot_ios-Info.plist:
```diff
@@ -59,5 +59,7 @@
 	$additional_plist_content
 	$plist_launch_screen_name
 	<key>CADisableMinimumFrameDurationOnPhone</key><true/>
+	<key>LaunchScreenImageMode</key>
+	<string>$launch_screen_image_mode</string>
```
No sure if adding it to `.plist` is the best way, it might interfere with system setting or App Store submission rules (at least at some point, unsupported properties in `Info.plist` was sufficient to fail provisioning), project setting probably is a more appropriate place for it. Exporter can be easily modified to store extra setting (it's already used for remaps):

```diff
diff --git a/editor/export/editor_export_platform.cpp b/editor/export/editor_export_platform.cpp
index f8358dd072..55c58a10cb 100644
--- a/editor/export/editor_export_platform.cpp
+++ b/editor/export/editor_export_platform.cpp
@@ -1525,7 +1525,7 @@ Error EditorExportPlatform::export_project_files(const Ref<EditorExportPreset> &
 		custom_list.append_array(export_plugins[i]->_get_export_features(Ref<EditorExportPlatform>(this), p_debug));
 	}
 
-	ProjectSettings::CustomMap custom_map;
+	ProjectSettings::CustomMap custom_map = get_custom_project_settings(p_preset);
 	if (path_remaps.size()) {
 		if (true) { //new remap mode, use always as it's friendlier with multiple .pck exports
 			for (int i = 0; i < path_remaps.size(); i += 2) {
diff --git a/editor/export/editor_export_platform.h b/editor/export/editor_export_platform.h
index ed49451433..37cfaac5cb 100644
--- a/editor/export/editor_export_platform.h
+++ b/editor/export/editor_export_platform.h
@@ -336,6 +336,7 @@ public:
 	virtual void get_platform_features(List<String> *r_features) const = 0;
 	virtual void resolve_platform_feature_priorities(const Ref<EditorExportPreset> &p_preset, HashSet<String> &p_features) {}
 	virtual String get_debug_protocol() const { return "tcp://"; }
+	virtual HashMap<String, Variant> get_custom_project_settings(const Ref<EditorExportPreset> &p_preset) const { return HashMap<String, Variant>(); }
 
 	EditorExportPlatform();
 };
diff --git a/platform/ios/export/export_plugin.cpp b/platform/ios/export/export_plugin.cpp
index 370e7345be..f253e11ec5 100644
--- a/platform/ios/export/export_plugin.cpp
+++ b/platform/ios/export/export_plugin.cpp
@@ -198,6 +198,19 @@ static const DataCollectionInfo data_collect_purpose_info[] = {
 	{ "Other", "NSPrivacyCollectedDataTypePurposeOther" },
 };
 
+static const String export_method_string[] = {
+	"app-store",
+	"development",
+	"ad-hoc",
+	"enterprise"
+};
+static const String storyboard_image_scale_mode[] = {
+	"center",
+	"scaleAspectFit",
+	"scaleAspectFill",
+	"scaleToFill"
+};
+
 String EditorExportPlatformIOS::get_export_option_warning(const EditorExportPreset *p_preset, const StringName &p_name) const {
 	if (p_preset) {
 		if (p_name == "application/app_store_team_id") {
@@ -402,19 +415,28 @@ void EditorExportPlatformIOS::get_export_options(List<ExportOption> *r_options)
 	r_options->push_back(ExportOption(PropertyInfo(Variant::COLOR, "storyboard/custom_bg_color"), Color()));
 }
 
+HashMap<String, Variant> EditorExportPlatformIOS::get_custom_project_settings(const Ref<EditorExportPreset> &p_preset) const {
+	HashMap<String, Variant> settings;
+
+	int image_scale_mode = p_preset->get("storyboard/image_scale_mode");
+	String value;
+
+	switch (image_scale_mode) {
+		case 0: {
+			String logo_path = GLOBAL_GET("application/boot_splash/image");
+			bool is_on = GLOBAL_GET("application/boot_splash/fullsize");
+			// If custom logo is not specified, Godot does not scale default one, so we should do the same.
+			value = (is_on && logo_path.length() > 0) ? "scaleAspectFit" : "center";
+		} break;
+		default: {
+			value = storyboard_image_scale_mode[image_scale_mode - 1];
+		}
+	}
+	settings["ios/launch_screen_image_mode"] = value;
+	return settings;
+}
+
 void EditorExportPlatformIOS::_fix_config_file(const Ref<EditorExportPreset> &p_preset, Vector<uint8_t> &pfile, const IOSConfigData &p_config, bool p_debug) {
-	static const String export_method_string[] = {
-		"app-store",
-		"development",
-		"ad-hoc",
-		"enterprise"
-	};
-	static const String storyboard_image_scale_mode[] = {
-		"center",
-		"scaleAspectFit",
-		"scaleAspectFill",
-		"scaleToFill"
-	};
 	String dbg_sign_id = p_preset->get("application/code_sign_identity_debug").operator String().is_empty() ? "iPhone Developer" : p_preset->get("application/code_sign_identity_debug");
 	String rel_sign_id = p_preset->get("application/code_sign_identity_release").operator String().is_empty() ? "iPhone Distribution" : p_preset->get("application/code_sign_identity_release");
 	bool dbg_manual = !p_preset->get_or_env("application/provisioning_profile_uuid_debug", ENV_IOS_PROFILE_UUID_DEBUG).operator String().is_empty() || (dbg_sign_id != "iPhone Developer" && dbg_sign_id != "iPhone Distribution");
diff --git a/platform/ios/export/export_plugin.h b/platform/ios/export/export_plugin.h
index e18e2c5ed6..8fc436153a 100644
--- a/platform/ios/export/export_plugin.h
+++ b/platform/ios/export/export_plugin.h
@@ -201,6 +201,8 @@ public:
 		return list;
 	}
 
+	virtual HashMap<String, Variant> get_custom_project_settings(const Ref<EditorExportPreset> &p_preset) const override;
+
 	virtual Error export_project(const Ref<EditorExportPreset> &p_preset, bool p_debug, const String &p_path, BitField<EditorExportPlatform::DebugFlags> p_flags = 0) override;
 
 	virtual bool has_valid_export_configuration(const Ref<EditorExportPreset> &p_preset, String &r_error, bool &r_missing_templates, bool p_debug = false) const override;
diff --git a/platform/ios/os_ios.h b/platform/ios/os_ios.h
index b25bbaa358..e9cbfd216c 100644
--- a/platform/ios/os_ios.h
+++ b/platform/ios/os_ios.h
@@ -133,7 +133,7 @@ public:
 	void on_enter_background();
 	void on_exit_background();
 
-	virtual Rect2 calculate_boot_screen_rect(const Size2 window_size, const Size2 imgrect_size) const override;
+	virtual Rect2 calculate_boot_screen_rect(const Size2 &p_window_size, const Size2 &p_imgrect_size) const override;
 };
 
 #endif // IOS_ENABLED
diff --git a/platform/ios/os_ios.mm b/platform/ios/os_ios.mm
index 99f57803e4..40e322a64f 100644
--- a/platform/ios/os_ios.mm
+++ b/platform/ios/os_ios.mm
@@ -660,20 +660,20 @@ void OS_IOS::on_exit_background() {
 	}
 }
 
-Rect2 OS_IOS::calculate_boot_screen_rect(const Size2 window_size, const Size2 imgrect_size) const {
-	NSString *scalemodestr = [[NSBundle mainBundle] objectForInfoDictionaryKey:@"LaunchScreenImageMode"];
-
-	if ([scalemodestr isEqualToString:@"scaleAspectFit"]) {
-		return window_size.fit_keep_aspect_centered(imgrect_size);
-	} else if ([scalemodestr isEqualToString:@"scaleAspectFill"]) {
-		return window_size.fit_keep_aspect_covered(imgrect_size);
-	} else if ([scalemodestr isEqualToString:@"scaleToFill"]) {
-		return Rect2(Point2(), window_size);
-	} else if ([scalemodestr isEqualToString:@"center"]) {
-		return this->OS_Unix::calculate_boot_screen_rect(window_size, imgrect_size);
+Rect2 OS_IOS::calculate_boot_screen_rect(const Size2 &p_window_size, const Size2 &p_imgrect_size) const {
+	String scalemodestr = GLOBAL_GET("ios/launch_screen_image_mode");
+
+	if (scalemodestr == "scaleAspectFit") {
+		return p_window_size.fit_keep_aspect_centered(p_imgrect_size);
+	} else if (scalemodestr == "scaleAspectFill") {
+		return p_window_size.fit_keep_aspect_covered(p_imgrect_size);
+	} else if (scalemodestr == "scaleToFill") {
+		return Rect2(Point2(), p_window_size);
+	} else if (scalemodestr == "center") {
+		return OS_Unix::calculate_boot_screen_rect(p_window_size, p_imgrect_size);
 	} else {
-		WARN_PRINT(vformat("Boot screen scale mode mismatch between iOS and Godot: %s not supported", scalemodestr.UTF8String));
-		return this->OS_Unix::calculate_boot_screen_rect(window_size, imgrect_size);
+		WARN_PRINT(vformat("Boot screen scale mode mismatch between iOS and Godot: %s not supported", scalemodestr));
+		return OS_Unix::calculate_boot_screen_rect(p_window_size, p_imgrect_size);
 	}
 }
 
```

bruvzg commented on platform/ios/os_ios.h:
```diff
@@ -132,6 +132,8 @@ class OS_IOS : public OS_Unix {
 	void on_enter_background();
 	void on_exit_background();
+
+	virtual Rect2 calculate_boot_screen_rect(const Size2 window_size, const Size2 imgrect_size) const override;
```
```suggestion
	virtual Rect2 calculate_boot_screen_rect(const Size2 &p_window_size, const Size2 &p_imgrect_size) const override;
```

bruvzg commented on core/math/vector2.cpp:
```diff
@@ -209,3 +210,43 @@ Vector2::operator String() const {
 }
+
+Rect2 Vector2::fit_keep_aspect_centered(const Vector2 &p_rect) const {
+	const real_t available_ratio = width / height;
+	const real_t fit_ratio = p_rect.width / p_rect.height;
```
```suggestion
	real_t available_ratio = width / height;
	real_t fit_ratio = p_rect.width / p_rect.height;
```

Usually we do not use `const` for primitive types, there's no difference for compiler, and it's just adding visual clutter.

bruvzg commented on core/os/os.cpp:
```diff
@@ -222,6 +222,23 @@ uint64_t OS::get_embedded_pck_offset() const {
+Rect2 OS::calculate_boot_screen_rect(const Size2 window_size, const Size2 imgrect_size) const {
+	Rect2 screenrect;
+	if (window_size.width > window_size.height) {
+		//scale horizontally
```
```suggestion
		// Scale horizontally.
```

Same for the rest of comments.

bruvzg reviewed (commented):
Seems to be working as expected.

jkirsteins commented:
> Please try to avoid pushing so many times in a row (10 times in 35 minutes), it really puts pressure on the automatic checks

Apologies, I didn't realize. Will avoid that.

akien-mga commented:
The additions to Vector2 and OS seem highly specific and not really warranted to be core additions.

Please add such code as helper methods where it's needed and not in core - see https://docs.godotengine.org/en/stable/contributing/development/best_practices_for_engine_contributors.html#prefer-local-solutions

akien-mga reviewed (changes requested):
I don't see (a priori) a reason to add a new API to Vector2 to calculate some boot splash metrics.

In second review, I see the OS one is used in RendererRD so I guess that's needed, though that still feels a bit weird architecture wise.

jkirsteins commented:
Thanks for the prompt feedback everyone.

> I don't see (a priori) a reason to add a new API to Vector2 to calculate some boot splash metrics.

I'll move the methods. But just to share my thought process - I chose adding them to the core, because the functionality itself (fitting a size inside another size given a stretch mode) is not specific to iOS. In fact, this code already exists (it is essentially what `TextureRect` does with its stretch modes). But since that code is located in `TextureRect`, I couldn't cleanly reuse it.

So my rationale was that by putting this in core, it could help unify this type of aspect fitting logic in the future.

However, I see how my approach does not fully align with existing conventions, so I'm happy to move it.

> In second review, I see the OS one is used in RendererRD so I guess that's needed, though that still feels a bit weird architecture wise.

An obvious other solution could be to have an OS-specific branch in the RendererRD, but I think it is also not ideal (there is no precedent to branch like that in this class). However, as it's a small change specific to just one platform, maybe that's preferable, instead of complicating the OS API. 

Please let me know if you prefer me to change the singleton approach as well.

**Next steps**: 
- I'll move the new Vector2 methods to the iOS singleton later today, and 
- I'll keep the OS singleton approach for now, unless there's an explicit ask to change that.

akien-mga commented:
> **Next steps**:
> 
> * I'll move the new Vector2 methods to the iOS singleton later today, and
> * I'll keep the OS singleton approach for now, unless there's an explicit ask to change that.

Sounds good to me!

> I'll move the methods. But just to share my thought process - I chose adding them to the core, because the functionality itself (fitting a size inside another size given a stretch mode) is not specific to iOS. In fact, this code already exists (it is essentially what `TextureRect` does with its stretch modes). But since that code is located in `TextureRect`, I couldn't cleanly reuse it.

That's a good point. Our current guidelines would likely be to hardcode it again in iOS for now, but look later if we want to refactor and harmonize this code with a core API. We often do that to first confirm that there's a frequent need for a bit of API before solidifying it in the core.

jkirsteins commented:
Pushed an update that moves the 2 methods to iOS. I also removed the tests I had added to Vector2 (they were useful to validate everything while it was WIP, but not sure they would add a lot of value in the iOS singleton).

I already squashed it, but in retrospect I realize that makes it harder to see what I changed. I'll wait for the review to be finalized next time before squashing, sorry about the inconvenience.

akien-mga reviewed (approved):
Looks good to me!

Repiteo commented:
Thanks!
